proof/refine/ARM/Schedule_R.thy

```diff
@@ -3368,5 +3317,840 @@ lemma possible_switch_to_corres:
+               in whileLoop_terminates_inv)
+    apply simp
+   apply clarsimp
+   apply (rule awaken_body_release_queue_length)
+  apply (rule release_queue_length_well_founded)
+  done
+
+lemma get_sc_refill_ready_no_fail[wp]:
+  "no_fail (\<lambda>s. pred_map \<top> (scs_of s) sc_ptr) (get_sc_refill_ready sc_ptr)"
+  apply (clarsimp simp: get_sc_refill_ready_def)
+  apply (wpsimp wp: get_sched_context_no_fail_stronger)
+  apply (clarsimp simp: vs_all_heap_simps)
+  done
+
+lemma get_tcb_refill_ready_no_fail:
+  "no_fail (\<lambda>s. pred_map2' \<top> (tcb_scps_of s) (sc_refill_cfgs_of s) t) (get_tcb_refill_ready t)"
+  apply (clarsimp simp: get_tcb_refill_ready_def get_tcb_obj_ref_def thread_get_def)
+  apply wpsimp
+  apply (clarsimp simp: vs_all_heap_simps get_tcb_def)
+  done
+
+lemma release_q_non_empty_and_ready_no_fail:
+  "no_fail valid_release_q release_q_non_empty_and_ready"
+  apply (clarsimp simp: release_q_non_empty_and_ready_def)
+  apply (wpsimp wp: get_tcb_refill_ready_no_fail)
+  apply (fastforce simp: valid_release_q_def obj_at_kh_kheap_simps vs_all_heap_simps)
+  done
+
+lemma reprogram_timer_update_release_queue_update_monad_commute:
+  "monad_commute \<top> (modify (reprogram_timer_update (\<lambda>_. b))) (modify (release_queue_update q))"
+  apply (clarsimp simp: monad_commute_def modify_def get_def put_def bind_def return_def)
+  done
+
+(* FIXME RT: move to NonDetMonadLemmaBucket, next to monad_commute_simple?*)
+lemma monad_commute_simple':
+  "monad_commute \<top> a b \<Longrightarrow> (do x \<leftarrow> a; y \<leftarrow> b; g x y od) = (do y \<leftarrow> b; x \<leftarrow> a; g x y od)"
+  apply (clarsimp simp: monad_commute_def)
+  apply (fastforce simp: bind_def' return_def)
+  done
+
+(* FIXME RT: move?*)
+lemma bind_return_noop:
+  "do y \<leftarrow> a; b od = do y \<leftarrow> a; z \<leftarrow> b; return () od"
+  by simp
+
+(* FIXME RT: move?*)
+lemma filter_hd_equals_tl:
+  "\<lbrakk>distinct q; q \<noteq> []\<rbrakk> \<Longrightarrow> filter ((\<noteq>) (hd q)) q = tl q"
+  apply (induct q rule: length_induct)
+  apply (rename_tac list)
+  apply (case_tac list; simp)
+  apply (fastforce simp: filter_id_conv)
+  done
+
+lemma release_queue_modify_tl:
+  "\<lbrakk>(s, s') \<in> state_relation; valid_release_q s; release_queue s \<noteq> []\<rbrakk>
+   \<Longrightarrow> (s\<lparr>release_queue := filter ((\<noteq>) (hd (release_queue s))) (release_queue s)\<rparr>,
+        s'\<lparr>ksReleaseQueue := tl (release_queue s)\<rparr>) \<in> state_relation"
+  apply (clarsimp simp: state_relation_def cdt_relation_def)
+  apply (clarsimp simp: release_queue_relation_def)
+  apply (prop_tac "ksReleaseQueue s' \<noteq> []", fastforce)
+  apply (rule filter_hd_equals_tl)
+   apply (prop_tac "distinct (release_queue s)")
+    apply (clarsimp simp: valid_release_q_def)
+   apply simp+
+  done
+
+(* FIXME RT: move? Is this useful?*)
```

Comment:
Do you have any specific use for it? Otherwise it does look a bit too specialised.